### DISCUSSION TITLE
Conditional Expression and Null check

### DISCUSSION THREAD

**Discussion by HemalathaGowdar** [mapstruct/mapstruct#2785] [Q&A]
When am giving conditional Expression , the NULL check (Null value check strategy) is getting replaced by conditional expression method. I need NULL check too irrespective of conditionalExpression. Could u please suggest

**Comment by Zegveld** [comment]
You want different behavior for null-check and conditional result?

For example:
```
if (source.getField() != null) {
  if (condition(source.getField())) {
    target.setField(source.getField());
  } else {
    // do something or nothing because condition did not match
  }
} else {
  // do something or nothing because source was null
}
```

I'm curious as to why the different behavior is wanted and what the difference would be.
There might be another solution to this besides using conditional expressions.

**Reply by filiphr** [reply]
@HemalathaGowdar I am curious to understand why you are not doing the null check in your conditional expression. There are valid usecases where you would treat `null` as a present value.

**Reply by HemalathaGowdar** [reply]
@filiphr :  I am invoking my new method to just check whether this attribute name is present in the consumer list to set  or not . This I have explained in Access to the target property name #2688. I am not passing the value of the attribute to do null check in the conditionalExpression method , so when it is generating , its failing with nullpointerexception in case of null values.

Ex:	@Mapping(source ="secName" , target ="sName" , conditionExpression = "java(variablePresenceCheck(\"sName\"))" )

**Reply by HemalathaGowdar** [reply]
> You want different behavior for null-check and conditional result?
> 
> For example:
> 
> ```
> if (source.getField() != null) {
>   if (condition(source.getField())) {
>     target.setField(source.getField());
>   } else {
>     // do something or nothing because condition did not match
>   }
> } else {
>   // do something or nothing because source was null
> }
> ```
> 
> I'm curious as to why the different behavior is wanted and what the difference would be. There might be another solution to this besides using conditional expressions.

when it generates  , am expecting like below ..
if (source.getField() != null) {
  if (conditionalExpression (source.getFieldName())) {

But if (source.getField() != null) { -- this is not happening , when we are converting the fields to diff data type .its failing with Null pointer

**Reply by Zegveld** [reply]
Seeing as you do not want different behavior, couldn't you do the following until #2688 is done?
`@Mapping(source ="secName" , target ="sName" , conditionExpression = "java(variablePresenceCheck(source.getSecName(), "sName"))" )`
and have a `variablePresenceCheck` that also does the null check with an `Object` as first parameter.
For example:
```
boolean variablePresenceCheck(Object field, String targetFieldName) {
    return field != null && variablePresenceCheck(targetFieldName);
}
```
After #2688 this is most likely also the way you want to implement the generic presence check.
